src/internal.rs

```diff
@@ -527,23 +543,18 @@ pub fn recover(
     })
     .collect();
 
-  // we also ensure that no message parsing errors occurred, we should
-  // recover with only a threshold number of messages, so any error that
-  // occurs should be fatal for recovery
-  if splits.iter().any(|x| x.is_err()) {
-    return Err(Error::MessageParse);
-  }
-
   // check that decrypted measurements all have the same value.
   // we can unwrap here because errors should have been caught above
-  let measurement = &splits[0].as_ref().unwrap().0;
-  for to_chk in splits.iter().skip(1) {
-    let measurement_to_chk = &to_chk.as_ref().unwrap().0;
-    if measurement_to_chk != measurement {
-      return Err(Error::ClientMeasurementMismatch(
-        serde_json::to_string(measurement).unwrap(),
-        serde_json::to_string(&measurement_to_chk).unwrap(),
-      ));
+  if let Some(measurement) = &splits.iter().find(|res| res.is_ok()) {
+    let measurement = &measurement.as_ref().unwrap().0;
+    for to_chk in splits.iter().flatten() {
+      let measurement_to_chk = &to_chk.0;
+      if measurement_to_chk != measurement {
+        return Err(Error::ClientMeasurementMismatch(
+          serde_json::to_string(measurement).unwrap(),
+          serde_json::to_string(&measurement_to_chk).unwrap(),
+        ));
```

Comment:
Would it make sense to stuff the error back into the `splits` vec, so this could return one less layer of `Result`?